# Romanian Models Evaluation and Interpretability Metrics

This notebook exposes the results of the classification models trained and evaluated on Romanian datasets.
We compare four setups across Sentiment Analysis and Emotion Detection tasks:

**Sentiment Analysis:**
- **BERT_Sentiment_Articles**

**Emotion Detection (RED dataset):**
- **RoBERT_Emotion_RED**
- **RoBERT_Conv_Emotion_RED**
- **RoBERT_Ensemble_Emotion_RED**

## 1. Metrics Comparison
Since the tasks differ, we present the Sentiment Analysis metrics (Accuracy, F1) and Emotion Detection metrics (Micro/Macro F1) separately.

In [1]:
import pandas as pd
import json

# 1. Sentiment Analysis
with open('1_BERT_Sentiment_Articles/Results/bert_romanian_metrics.json') as f:
    sa_metrics = json.load(f)

df_sa = pd.DataFrame([sa_metrics], index=['BERT_Sentiment_Articles'])
print("--- Sentiment Analysis Metrics ---")
display(df_sa.round(4))

# 2. Emotion Detection (RED)
emo_paths = {
    'RoBERT_Emotion': '2_RoBERT_Emotion_RED/Results/robert_red_metrics.json',
    'RoBERT_Conv_Emotion': '3_RoBERT_Conv_Emotion_RED/Results/robert_conv_metrics.json',
    'RoBERT_Ensemble_Emotion': '4_RoBERT_Ensemble_Emotion_RED/Results/robert_ensemble_metrics.json'
}

emo_data = []
for name, path in emo_paths.items():
    with open(path) as f:
        data = json.load(f)
        emo_data.append({
            'Model': name,
            'Micro F1': data.get('micro_f1', 0),
            'Macro F1': data.get('macro_f1', 0),
            'Hamming Loss': data.get('hamming_loss', 0)
        })
        
df_emo = pd.DataFrame(emo_data).set_index('Model')
print("\n--- Emotion Detection Metrics ---")
display(df_emo.round(4))

--- Sentiment Analysis Metrics ---


,accuracy,precision,recall,f1
BERT_Sentiment_Articles,0.7287,0.7273,0.7326,0.7296



--- Emotion Detection Metrics ---


,Micro F1,Macro F1,Hamming Loss
Model,,,
RoBERT_Emotion,0.6790,0.6768,0.1096
RoBERT_Conv_Emotion,0.6681,0.6663,0.1086
RoBERT_Ensemble_Emotion,0.6909,0.6877,0.1072


## 2. First Sample Analysis - Interpretability (LIME)
Below we expose the LIME interpretability for the first sample (`example_000.html`) for all models.
Since these are interactive HTML files, we render them directly in the notebook using `IFrame`.

In [2]:
from IPython.display import IFrame, display, HTML

display(HTML("<h3>BERT_Sentiment_Articles</h3>"))
display(IFrame(src='./1_BERT_Sentiment_Articles/Results/lime_explanations/example_000.html', width='100%', height=400))

display(HTML("<hr><h3>RoBERT_Emotion_RED</h3>"))
display(IFrame(src='./2_RoBERT_Emotion_RED/Results/lime_explanations/example_000.html', width='100%', height=400))

display(HTML("<hr><h3>RoBERT_Conv_Emotion_RED</h3>"))
display(IFrame(src='./3_RoBERT_Conv_Emotion_RED/Results/lime_explanations/example_000.html', width='100%', height=400))

display(HTML("<hr><h3>RoBERT_Ensemble_Emotion_RED</h3>"))
display(IFrame(src='./4_RoBERT_Ensemble_Emotion_RED/Results/lime_explanations/example_000.html', width='100%', height=400))

### Description and Interpretation

**Metrics Interpretation:**
- **Sentiment Analysis:** The BERT model on Romanian Articles achieves an accuracy and F1 score of ~72.9%.
- **Emotion Detection:** The models perform consistently, with the **Ensemble** model performing best in terms of Micro F1 (~69.1%) and Macro F1 (~68.8%), followed closely by the standard RoBERT (~67.9% Micro F1). The Convolutional RoBERT comes slightly behind (~66.8% Micro F1). This indicates that ensembling effectively boosts Emotion classification performance on the RED dataset.

**Sample Interpretability (LIME):**
- The interactive LIME frames above allow you to see the exact tokens driving the classification of the first sample.
- For the multi-label Emotion Detection task, LIME typically shows the explanations for the top predicted classes, highlighting which words contributed positively or negatively to predicting specific emotions like *sadness*, *joy*, or *anger*.